client/verta/verta/runtime.py

```diff
@@ -0,0 +1,237 @@
+    Raise an error on failure.
+    """
+    try:
+        return json.dumps(value)
+    except TypeError as type_err:
+        print("provided data type cannot be converted to JSON for logging.")
+        raise type_err
+
+
+def _validate_s3(
+        value: str,
+        pattern: re.Pattern = _S3_REGEX
+    ) -> None:
+    """
+    Check the provided value for any special characters that would
+    violate Amazon S3 object key naming rules. Raise ValueError if any
+    are found.
+    """
+    if len(value) > 100:
+        raise ValueError(" provided key value must be 100 characters or less in length.")
+    if not pattern.match(value):
+        raise ValueError(f" provided value \"{value}\" contains non-alphanumeric "
+                         f"characters. (dashes and underscores permitted)")
+
+
+def log(key: str, value: Any) -> None:
+    """
+    Updates current logging context dict with provided key and value.
+    For use within a model's :meth:`~verta.registry.VertaModelBase.predict`
+    method to collect logging context.
+
+    .. note::
+        Multithreading of calls to log() within a model's predict() method is not
+        currently supported.
+
+    Parameters
+    ----------
+    key : str
+        String value to use as data label (key) in logging context, with a limit of
+        100 characters or less.
+    value : Any
+        Any `JSON serializable <https://docs.python.org/3/library/json.html#json.JSONEncoder>`__
+        value you wish to include in the context logs.
+    Returns
+    -------
+    None
+
+    Raises
+    -------
+    TypeError
+        If `validate` was set to ``True`` on the active :class:`context`, and
+        `value` is not a JSON-serializable type.
+    ValueError
+        If `key` provided contains non-alphanumeric characters.  Dashes `-`
+        and underscores `_` are permitted.
+
+    Examples
+    --------
+    .. code-block:: python
+       :emphasize-lines: 11,14
+
+        from verta import runtime
+
+        # Sample model code:
+        class MyModel(VertaModelBase):
+            def __init__(self, artifacts):
+                pass
+
+            @verify.io
+            def predict(self, data):
+                logs = {'some_stuff': 'I_care_about'}
+                runtime.log('my_logging_key', logs)
+                prediction_output = 'some_output'
+                more_logs = ['a', 'list', 'of', 'things']
+                runtime.log('things', more_logs)
+                return prediction_output
+
+    """
+    if _get_validate_flag():
+        _validate_json(value)
+    _validate_s3(key)
+    local_logs: Dict[str, Any] = _get_thread_logs()
+    local_logs.update({key: value})
+
+
+class context:
+    """
+    Context manager for aggregating key-value pairs into a custom log entry.
+    It should not be necessary to instantiate this class directly unless you
+    wish to export log entries manually to a resource other than Amazon S3.
```

Comment:
I wouldn't advise mentioning that at this time. It implicitly puts us on the hook to support users who wish to export to other object stores, which is not in-scope for this feature.

```suggestion
    It should not be necessary to instantiate this class directly.
```

Reply:
Our design doc also lists "user debugging their own model logging calls" as a use case.